chainer/link.py

```diff
@@ -130,15 +133,21 @@ def add_param(self, name, shape, dtype=numpy.float32):
                 name will be removed.
             shape (int or tuple of ints): Shape of the parameter array.
             dtype: Data type of the parameter array.
+            initializer(chainer.initializer.Initializer): If not ``None``,
+                the data is initialized by initializer. In that case,
+                ``dtype`` argument is ignored.
 
         """
         d = self.__dict__
         if name in d:
             raise AttributeError(
                 'cannot register a new parameter %s: attribute exists'
                 % name)
-        data = self.xp.full(shape, numpy.nan, dtype=dtype)
-        grad = data.copy()
+        if initializer is None:
+            data = self.xp.full(shape, numpy.nan, dtype=dtype)
+        else:
+            data = initializers.generate_array(initializer, shape, self.xp)
+        grad = self.xp.full_like(data, numpy.nan)
```

Comment:
It is well-known that filling the gradients with ``np.nan`` is important to avoid bugs, however, I also think that is slightly breaks backward compatibility when the users use the wrong implementation.
Is this modification ok for the chainer team?